docs/proposals/27-falco-extension.md

```diff
@@ -0,0 +1,318 @@
+- Forced updates always update to the latest supported version.
+- There may be multiple supported versions.
+
+#### Falco Rules
+
+The Gardener extension will deploy matching rules for the Falco release from
+[8]. `falco_rules.yaml` will be deployed by default and Gardener will bring 
+extensions to the rule set to remediate false positives if necessary. 
+`falco-incubating_rules.yaml` and `falco-sandbox_rules.yaml` will only be 
+deployed if selected and there will be no exceptions by Gardener for 
+false positve events.
+
+#### Falcoctl
+
+Users with many clusters requested support for falcoctl [9]. There is 
+extra effort for configuration as an external repository for rules files
+and plug-ins is required for the benefit of controlling rules and 
+plugins centrally. We plan to pass the falcoctl configuration and disable 
+all Gardener managed rules for this use case.
+
+#### Falco Event Ingestor
+
+The Falco event ingestor receives Falco events from Falcosidekick, verifies
+the JWT token as well as the event body and stores it into an event store
+(currently a Postgres database). Due to receiving events from potentially 
+hundreds of clusters the Falco event ingestor must be highly scalable as
+well as to provide additional hardening including the following checks:
+
+- event rate limiting based on cluster identities
+- strict validation of events
+
+The implementation of the event store is deliberately not specified in this
+enhancement proposal. There are however several requirements toward the
+event store:
+
+- the event store must be able to store events generated from a whole Gardener 
+  landscape with potentially thousands of clusters.
+- events must be kept for 7 months
+- the events store must be highly available
+- a backup/restore implementation is required
+
+#### Gardener Dashboard
+
+Research and prototypeing work is ongoing. Our current working assumption is
+to add a tab to the shoot cluster details page to visualize events. We also
+consider adding the event page to the project overview and allow users to 
+filter by cluster.
+
+#### Alerting
+
+Research is ongoing.
+
+## Alternatives
+
+Our initial focus was to protect the Gardener infrastructure by deploying
+Falco to seed clusters directly using a `ControllerDeployment` containing
+a Falco helm chart [2]. We deployed a central Loki log aggregation system [3]
+to store Falco events and used Grafana [4] for visualization. While this
+provided great insights on what is happening in the Gardener infrastructure
+it was not optimal in several ways:
+
+1. The solution was tailored towards the Gardener infrastructure which
+cannot be applied to user shoot clusters. Deploying Falco into user clusters
+demands a solution as outlined in this proposal.
+
+2. The Gardener ecosystem is licensed under the Apache 2.0 open-source license
+while Loki and Grafana are not. We therefore cannot rely on central components
+which must licensed separately by adopters.
```

Comment:
The approach described in the GEP only targets shoot clusters via the existing extension mechanics. What about the garden and soil clusters which may not be shoot clusters?
(The approach described in the alternatives section could handle seed clusters, but could not handle the garden cluster.)

Reply:
> The approach described in the GEP only targets shoot clusters via the existing extension mechanics.

Yes, it does. Soil- and garden clusters cannot be handled consistently using this approach. The original approach also did not cover this aspect.

This said, the solution for specialised set-ups is quite straightforward. As the configuration for a Gardener landscape is quite static the landscape maintenance can deploy- and update helm charts to mimic the behaviour of the extension. We have shows this to work somewhat with the original approach. The biggest problem were diverging Falco rules because of different approaches to set up Kubernetes. Also the official Falco rules contain some fine-tuning for GKE.

I would therefore omit this topic from this proposal.

Reply:
As there are currently discussions ongoing regarding extension support for the garden runtime cluster, i.e. the gardener-operator scenario, maybe there is a standard way coming soon.

/cc @kon-angelo

That being said, it is fine to keep the scope limited. Maybe just add this deliberate limitation in scope also to the non-goals section to make it more obvious.